Some code incorrectly compares Document objects using CSSParserContext's comparison operator

cdumez opened the pull request:
#### 1dc257f3194c1771300e2b81f8edce1f41cd7e62
<pre>
Some code incorrectly compares Document objects using CSSParserContext&apos;s comparison operator
<a href="https://bugs.webkit.org/show_bug.cgi?id=261044">https://bugs.webkit.org/show_bug.cgi?id=261044</a>

Reviewed by Ryosuke Niwa and Darin Adler.

Use a named operator==() in CSSParserContext to avoid so that can no longer
happen and fix code doing bad comparisons.

* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::operator==): Deleted.
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::Editor::replaceNodeFromPasteboard):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::checkPopoverValidity):
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::reportViolation const):

Canonical link: <a href="https://commits.webkit.org/267577@main">https://commits.webkit.org/267577@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a3a5cd432293057de621b30f91ffaf9afbc41dc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17049 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17372 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17841 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18833 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15949 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17239 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20607 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17512 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18169 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17250 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17594 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14777 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19647 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14836 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15470 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22180 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15824 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15637 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19966 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16236 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13764 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15394 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4070 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19757 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16068 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/parser/CSSParserContext.h:
```diff
@@ -101,8 +101,8 @@ struct CSSParserContext {
 
-    CSSParserContext(CSSParserMode, const URL& baseURL = URL());
-    WEBCORE_EXPORT CSSParserContext(const Document&, const URL& baseURL = URL(), const String& charset = emptyString());
+    explicit CSSParserContext(CSSParserMode, const URL& baseURL = URL());
```
Not sure this change is as valuable as the `Document` one. Seems unlikely this implicit constructor does harm anywhere.

cdumez replied:
I thought it was our policy in WebKit to mark constructors taking a single parameters (or single mandatory one in this case) as explicit. These 2 didn't follow our pattern and one of them caused this subtle bug as a result. This is why it was my personal preference to update both.

darinadler replied:
Well, we mark constructors that aren't just lossless type conversions as explicit. I think it's not perfectly clear that this is one of those cases.

darinadler reviewed (approved):
Another approach would be to delete the `==` operator and use a named function for `CSSParserContext` equality checks; not even sure why it’s so helpful to have `==` for that class.

Seems a shame that all these call sites become less elegant. Can we mitigate that? Maybe the above idea? Maybe only make the dangerous `CSSParserContext` constructor explicit? Maybe overload some functions to take `const Document&` and create the parser context or maybe just take the document and not the parser context in some?

cdumez committed:
Some code incorrectly compares Document objects using CSSParserContext's comparison operator
https://bugs.webkit.org/show_bug.cgi?id=261044

Reviewed by Ryosuke Niwa and Darin Adler.

Use a named operator==() in CSSParserContext to avoid so that can no longer
happen and fix code doing bad comparisons.

* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::operator==): Deleted.
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::Editor::replaceNodeFromPasteboard):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::checkPopoverValidity):
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::reportViolation const):

Canonical link: https://commits.webkit.org/267577@main
